extensions/grafana-loki/PushGrafanaLokiREST.cpp

```diff
@@ -0,0 +1,390 @@
+    client_.setReadTimeout(read_timeout->getMilliseconds());
+  }
+}
+
+void PushGrafanaLokiREST::setAuthorization(const core::ProcessContext& context) {
+  if (auto username = context.getProperty(PushGrafanaLokiREST::Username)) {
+    auto password = context.getProperty(PushGrafanaLokiREST::Password);
+    if (!password) {
+      throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Username is set, but Password property is not!");
+    }
+    std::string auth = *username + ":" + *password;
+    auto base64_encoded_auth = utils::string::to_base64(auth);
+    client_.setRequestHeader("Authorization", "Basic " + base64_encoded_auth);
+  } else if (auto bearer_token_file = context.getProperty(PushGrafanaLokiREST::BearerTokenFile)) {
+    if (!std::filesystem::exists(*bearer_token_file) || !std::filesystem::is_regular_file(*bearer_token_file)) {
+      throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Bearer Token File is not a regular file!");
+    }
+    std::ifstream file(*bearer_token_file, std::ios::binary);
+    std::stringstream buffer;
+    buffer << file.rdbuf();
+    std::string bearer_token = utils::string::trim(buffer.str());
+    client_.setRequestHeader("Authorization", "Bearer " + bearer_token);
+  } else {
+    client_.setRequestHeader("Authorization", std::nullopt);
+  }
+}
+
+void PushGrafanaLokiREST::initializeHttpClient(core::ProcessContext& context) {
+  auto url = utils::getRequiredPropertyOrThrow<std::string>(context, Url.name);
+  if (url.empty()) {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Url property cannot be empty!");
+  }
+  if (utils::string::endsWith(url, "/")) {
+    url += "loki/api/v1/push";
+  } else {
+    url += "/loki/api/v1/push";
+  }
+  logger_->log_debug("PushGrafanaLokiREST push url is set to: {}", url);
+  client_.initialize(utils::HttpRequestMethod::POST, url, getSSLContextService(context));
+}
+
+void PushGrafanaLokiREST::onSchedule(core::ProcessContext& context, core::ProcessSessionFactory&) {
+  setUpStateManager(context);
+  initializeHttpClient(context);
+  client_.setContentType("application/json");
+  client_.setFollowRedirects(true);
+
+  setUpStreamLabels(context);
+
+  if (auto log_line_metadata_attributes = context.getProperty(LogLineMetadataAttributes)) {
+    log_line_metadata_attributes_ = utils::string::splitAndTrimRemovingEmpty(*log_line_metadata_attributes, ",");
+  }
+
+  auto tenant_id = context.getProperty(TenantID);
+  if (tenant_id && !tenant_id->empty()) {
+    client_.setRequestHeader("X-Scope-OrgID", tenant_id);
+  } else {
+    client_.setRequestHeader("X-Scope-OrgID", std::nullopt);
+  }
+  auto log_line_batch_wait = context.getProperty<core::TimePeriodValue>(LogLineBatchWait);
+  auto log_line_batch_size = context.getProperty<uint64_t>(LogLineBatchSize);
+  if (log_line_batch_size && *log_line_batch_size < 1) {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Log Line Batch Size property is invalid!");
```

Comment:
Error messages should help the user fixing the problem when possible.
```suggestion
    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Log Line Batch Size property is missing or less than 1!");
```

Reply:
Good point, updated in 0f1a614f803abf8ea538df7f95330a953ae2afd7